tmail-backend/blob/secondary-blob-store/src/main/java/com/linagora/tmail/blob/secondaryblobstore/SecondaryBlobStoreDAO.java

```diff
@@ -0,0 +1,116 @@
+        this.firstBlobStoreDAO = firstBlobStoreDAO;
+        this.secondBlobStoreDAO = secondBlobStoreDAO;
+    }
+
+    @Override
+    public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStoreIOException, ObjectNotFoundException {
+        try {
+            return firstBlobStoreDAO.read(bucketName, blobId);
+        } catch (Exception ex) {
+            LOGGER.error("Fail to read from the first blob store with bucket name {} and blobId {}. Use second blob store", bucketName.asString(), blobId.asString(), ex);
+            return secondBlobStoreDAO.read(bucketName, blobId);
+        }
+    }
+
+    @Override
+    public Mono<InputStream> readReactive(BucketName bucketName, BlobId blobId) {
+        return Mono.from(firstBlobStoreDAO.readReactive(bucketName, blobId))
+            .onErrorResume(ex -> {
+                LOGGER.error("Fail to read from the first blob store with bucket name {} and blobId {}. Use second blob store", bucketName.asString(), blobId.asString(), ex);
+                return Mono.from(secondBlobStoreDAO.readReactive(bucketName, blobId));
+            });
+    }
+
+    @Override
+    public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.from(firstBlobStoreDAO.readBytes(bucketName, blobId))
+            .onErrorResume(ex -> {
+                LOGGER.error("Fail to read from the first blob store with bucket name {} and blobId {}. Use second blob store", bucketName.asString(), blobId.asString(), ex);
+                return Mono.from(secondBlobStoreDAO.readBytes(bucketName, blobId));
+            });
+    }
+
+    @Override
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, byte[] data) {
+        return Mono.from(firstBlobStoreDAO.save(bucketName, blobId, data))
+            .then(Mono.from(secondBlobStoreDAO.save(bucketName, blobId, data)));
+    }
+
+    @Override
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, InputStream inputStream) {
+        return Mono.fromCallable(() -> IOUtils.toByteArray(inputStream))
+            .flatMap(data -> Mono.from(firstBlobStoreDAO.save(bucketName, blobId, data))
+                .then((Mono.from(secondBlobStoreDAO.save(bucketName, blobId, data)))));
+    }
+
+    @Override
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, ByteSource content) {
+        return Mono.from(firstBlobStoreDAO.save(bucketName, blobId, content))
+            .then(Mono.from(secondBlobStoreDAO.save(bucketName, blobId, content)));
+    }
+
+    @Override
+    public Mono<Void> delete(BucketName bucketName, BlobId blobId) {
+        return Mono.from(firstBlobStoreDAO.delete(bucketName, blobId))
+            .then(Mono.from(secondBlobStoreDAO.delete(bucketName, blobId)));
+    }
+
+    @Override
+    public Mono<Void> delete(BucketName bucketName, Collection<BlobId> blobIds) {
+        return Mono.from(firstBlobStoreDAO.delete(bucketName, blobIds))
```

Comment:
Same thing than for `save` (we will have a queue for fixing partial failures and the blob won't be exposed by metadata so the only risk we have is residual metadata!